File: LeetCode/Explore/TopInterview/Blind75/String/424_longest_repeating_character_replacement.cpp
// two pointers sliding window hash table
// Complexity Analysis
// Time Complexity: O(|S|)O(∣S∣) where |S| represents the length of string SS. 
// In the worst case we might end up visiting every element of string SS twice, once by left pointer and once by right pointer. 
// Space Complexity: O(|S|)O(∣S∣). |S|∣S∣ when the window size is equal to the entire string SS. 

class Solution {
public:
  int characterReplacement(string s, int k) {
    unordered_map<char, int> mp;
    int count = 0, ans = 0;
    for (int r = 0, l = 0; r < s.size(); r++) {
      mp[s[r]]++;
      count = max(count, mp[s[r]]);
      // when repeating counts are too small to get one longest repeating string
      // sliding window shifts towards right 
      if (r - l + 1 - count > k) {
        mp[s[l]]--;
        l++;
      }
      ans = max(ans, r - l + 1);
    }
    return ans;
  }
};

File: LeetCode/Explore/TopInterview/Blind75/String/424_longest_repeating_character_replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
      maxLen, counter, l = 0, 0, 0
      mp = dict()
      
      for r, c in enumerate(s):
        mp[c] = mp.get(c, 0) + 1
        counter = max(counter, mp[c])
        if r - l + 1 - counter > k:
          mp[s[l]] -= 1 
          l += 1
        maxLen = max(maxLen, r - l + 1)
      
      return maxLen
